src/enumerable.cr

```diff
@@ -965,6 +965,41 @@ module Enumerable(T)
     ary
   end
 
+  private def qselect_internal(data : Array(T), left : Int, right : Int, k : Int)
+    loop do
+      return data[left] if left == right
+      pivot_index = left + (right-left)//2
+      pivot_index = qselect_partition_internal(data, left, right, pivot_index)
+      if pivot_index < 0
+        return nil
+      elsif k == pivot_index
+        return data[k]
+      elsif k < pivot_index
+        right = pivot_index - 1
+      else
+        left = pivot_index + 1
+      end
+    end
+  end
+
+  private def qselect_partition_internal(data : Array(T), left : Int, right : Int, pivot_index : Int) : Int
+    pivot_value = data[pivot_index]
+    data.swap(pivot_index, right)
+    store_index = left
+    (left..right).each do |i|
+      # if data[i] < pivot_value
+      cmp = data[i] <=> pivot_value
+      if cmp == -1
+        data.swap(store_index, i)
+        store_index += 1
+      elsif cmp.nil?
+        return -1
+      end
```

Comment:
Although most implementations of `#<=>` return -1 or 1, formally only comparisons against 0 are allowed:

```suggestion
      if cmp.nil?
        return -1
      elsif cmp < 0
        data.swap(store_index, i)
        store_index += 1
      end
```

The reordering is necessary so that the compiler doesn't try to compare nil against an integer.